File: Practice/programmers_077.py
def check(p):
    u = []
    try:
        for i in p:
            if i == '(':
                u.append('(')
            else:
                u.pop()
        return True
    except:
        return False

def split(p):
    cnt = 0
    for i in range(len(p)):
        if p[i] == '(':
            cnt += 1
        else:
            cnt -= 1
        if cnt == 0:
            break
    return p[:i+1], p[i+1:]

def convert (u):
    temp = ''
    for i in u:
        if i == '(':
            temp += ')'
        else:
            temp += '('
    return temp

def solution(p):
    answer = ''
    while len(p) != 0:
        u, p = split(p)
        if check(u):
            answer += u
        else:
            answer += '(' + solution(p) + ')' + convert(u[1:-1])
            break

    return answer


# 간소화
def solution(p):
    if p == '': 
        return p
    r = True
    c = 0
    
    for i in range(len(p)):
        if p[i] == '(': 
            c -= 1
        else: 
            c += 1
            
        if c > 0: 
            r = False
        if c == 0:
            if r:
                return p[:i+1] + solution(p[i+1:])
            else:
                return '('+solution(p[i+1:])+')'+''.join(['(' if x==')' else ')' for x in p[1:i]])

File: JavaScript/programmers_077.js
// LV.2 후보키

function getComb(arr, num) {
  let result = [];
  if (num === 1) return arr.map((item) => [item].join(""));
  arr.forEach((fixed, index, arr) => {
    const rest = arr.slice(index + 1);
    const comb = getComb(rest, num - 1);
    const answer = comb.map((item) => [fixed, ...item].join(""));
    result.push(...answer);
  });
  return result;
}

function isUnique(tuple) {
  const tmp = tuple.map((item) => item.join(""));
  const set = new Set(tmp);

  return tmp.length === set.size ? true : false;
}

function solution(relation) {
  var answer = 0;
  let condition = [];
  let column = relation[0].length;
  let arr = [];

  for (let i = 0; i < column; i++) {
    arr.push(i);
  }

  for (let i = 1; i <= column; i++) {
    condition.push(...getComb(arr, i));
  }

  while (condition.length > 0) {
    const columns = condition.shift().split("");
    const tuple = relation.map((row) => columns.map((col) => row[col]));
    if (isUnique(tuple) === true) {
      answer += 1;

      const colComTmp = [];
      for (let i = 0; i < condition.length; i++) {
        columns.map((col) => {
          if (!condition[i].includes(col)) {
            colComTmp.push(condition[i]);
          }
        });
      }
      condition = colComTmp;
    } else {
      continue;
    }
  }

  return answer;
}
